File: kernel/api/src/mesh_glb.cpp
#include <occccad/kernel/mesh_glb.hpp>

#include <algorithm>
#include <array>
#include <cstring>
#include <limits>
#include <sstream>
#include <stdexcept>
#include <string>

namespace occccad::kernel {
namespace {

void append_u32(std::vector<uint8_t>& output, const uint32_t value) {
    output.push_back(static_cast<uint8_t>(value));
    output.push_back(static_cast<uint8_t>(value >> 8U));
    output.push_back(static_cast<uint8_t>(value >> 16U));
    output.push_back(static_cast<uint8_t>(value >> 24U));
}

void append_float(std::vector<uint8_t>& output, const float value) {
    static_assert(sizeof(float) == sizeof(uint32_t));
    uint32_t bits = 0;
    std::memcpy(&bits, &value, sizeof(bits));
    append_u32(output, bits);
}

}  // namespace

std::vector<uint8_t> make_glb(const TessellationResult& mesh,
                              const std::string& topology_ids_json) {
    if (mesh.vertices.empty() || mesh.triangles.empty()) {
        throw std::invalid_argument("cannot encode an empty mesh as GLB");
    }

    std::vector<uint8_t> binary;
    std::vector<std::string> views, accessors;
    auto accessor = [&](size_t offset, size_t count, const char* type, int component,
                        const std::string& bounds = "") {
        const auto id = accessors.size();
        views.push_back("{\"buffer\":0,\"byteOffset\":" + std::to_string(offset) +
                        ",\"byteLength\":" + std::to_string(binary.size() - offset) + "}");
        accessors.push_back("{\"bufferView\":" + std::to_string(id) + ",\"componentType\":" +
                            std::to_string(component) + ",\"count\":" + std::to_string(count) +
                            ",\"type\":\"" + type + "\"" + bounds + "}");
        return id;
    };
    auto points = [&](const auto& vertices) {
        size_t offset = binary.size();
        std::array<double, 3> lo{vertices[0].x, vertices[0].y, vertices[0].z}, hi = lo;
        for (const auto& v : vertices) {
            const double values[]{v.x, v.y, v.z};
            for (int i = 0; i < 3; ++i) {
                lo[i] = std::min(lo[i], values[i]);
                hi[i] = std::max(hi[i], values[i]);
                append_float(binary, static_cast<float>(values[i]));
            }
        }
        std::ostringstream bounds;
        bounds << ",\"min\":[" << lo[0] << ',' << lo[1] << ',' << lo[2] << "],\"max\":[" << hi[0]
               << ',' << hi[1] << ',' << hi[2] << ']';
        return accessor(offset, vertices.size(), "VEC3", 5126, bounds.str());
    };
    const auto positions = points(mesh.vertices);
    size_t offset = binary.size();
    for (const auto& t : mesh.triangles) {
        append_u32(binary, t.v0);
        append_u32(binary, t.v1);
        append_u32(binary, t.v2);
    }
    const auto indices = accessor(offset, mesh.triangles.size() * 3, "SCALAR", 5125);
    if (mesh.face_ids.size() != mesh.triangles.size())
        throw std::invalid_argument("GLB face mapping count mismatch");
    offset = binary.size();
    for (auto id : mesh.face_ids)
        append_u32(binary, id);
    const auto faces = accessor(offset, mesh.face_ids.size(), "SCALAR", 5125);
    std::ostringstream cad;
    cad << "{\"schemaVersion\":1,\"units\":\"mm\",\"coordinateSpace\":\"PART_LOCAL\",\"faceIds\":"
        << faces << ",\"edges\":[";
    bool first = true;
    for (const auto& edge : mesh.edges) {
        if (edge.points.empty())
            continue;
        if (!first)
            cad << ',';
        first = false;
        cad << "{\"localId\":" << edge.local_id << ",\"positions\":" << points(edge.points) << '}';
    }
    cad << "],\"vertices\":[";
    first = true;
    // Vertex coordinates are small and independent of tessellation vertices.
    for (const auto& vertex : mesh.topology_vertices) {
        if (!first)
            cad << ',';
        first = false;
        cad << "{\"localId\":" << vertex.local_id << ",\"point\":[" << vertex.point.x << ','
            << vertex.point.y << ',' << vertex.point.z << "]}";
    }
    cad << "],\"stableIds\":" << topology_ids_json << "}";
    std::ostringstream json_stream;
    json_stream << "{\"asset\":{\"version\":\"2.0\",\"generator\":\"occccad\"},\"extensionsUsed\":["
                   "\"OCCCCAD_cad\"],\"extensions\":{\"OCCCCAD_cad\":"
                << cad.str() << "},\"buffers\":[{\"byteLength\":" << binary.size()
                << "}],\"bufferViews\":[";
    for (size_t i = 0; i < views.size(); ++i) {
        if (i)
            json_stream << ',';
        json_stream << views[i];
    }
    json_stream << "],\"accessors\":[";
    for (size_t i = 0; i < accessors.size(); ++i) {
        if (i)
            json_stream << ',';
        json_stream << accessors[i];
    }
    json_stream << "],\"meshes\":[{\"primitives\":[{\"attributes\":{\"POSITION\":" << positions
                << "},\"indices\":" << indices
                << "}]}],\"nodes\":[{\"mesh\":0}],\"scenes\":[{\"nodes\":[0]}],\"scene\":0}";
    std::string json = json_stream.str();
    while ((json.size() % 4U) != 0U)
        json.push_back(' ');

    constexpr uint32_t kGlbHeaderSize = 12U;
    constexpr uint32_t kChunkHeaderSize = 8U;
    if (json.size() + binary.size() > std::numeric_limits<uint32_t>::max() - 28ULL)
        throw std::length_error("GLB exceeds its 32-bit container limit");
    const auto total_size = static_cast<uint32_t>(kGlbHeaderSize + kChunkHeaderSize + json.size() +
                                                  kChunkHeaderSize + binary.size());
    std::vector<uint8_t> output;
    output.reserve(total_size);
    append_u32(output, 0x46546c67U);
    append_u32(output, 2U);
    append_u32(output, total_size);
    append_u32(output, static_cast<uint32_t>(json.size()));
    append_u32(output, 0x4e4f534aU);
    output.insert(output.end(), json.begin(), json.end());
    append_u32(output, static_cast<uint32_t>(binary.size()));
    append_u32(output, 0x004e4942U);
    output.insert(output.end(), binary.begin(), binary.end());
    return output;
}

}  // namespace occccad::kernel

File: web/apps/cad/src/cad/visual/mesh-glb.ts
import type { MeshData, VisualizationManifest, Vec3 } from "../../types";
export type DecodedVisual = {
    mesh: MeshData;
    visualization?: VisualizationManifest;
};
const emptyMesh = (): MeshData => ({ vertices: [], triangles: [], faceIds: [], edges: [], topologyVertices: [], stableIds: {} });
export function decodeMeshGLB(buffer: ArrayBuffer): DecodedVisual {
    const data = new DataView(buffer);
    if (buffer.byteLength < 20 || data.getUint32(0, true) !== 0x46546c67 || data.getUint32(4, true) !== 2 || data.getUint32(8, true) !== buffer.byteLength)
        throw new Error("Invalid GLB header");
    let document: any, binary: DataView | undefined;
    for (let offset = 12; offset < buffer.byteLength;) {
        if (offset + 8 > buffer.byteLength)
            throw new Error("Invalid GLB chunk header");
        const length = data.getUint32(offset, true), kind = data.getUint32(offset + 4, true);
        offset += 8;
        if (length % 4 || offset + length > buffer.byteLength)
            throw new Error("Invalid GLB chunk length");
        if (kind === 0x4e4f534a)
            document = JSON.parse(new TextDecoder().decode(new Uint8Array(buffer, offset, length)));
        if (kind === 0x004e4942)
            binary = new DataView(buffer, offset, length);
        offset += length;
    }
    if (document?.asset?.version !== "2.0")
        throw new Error("Unsupported GLB asset");
    const mesh = emptyMesh(), visualization = document.extensions?.OCCCCAD_visualization;
    if (!document.meshes?.length)
        return { mesh, visualization };
    const cad = document.extensions?.OCCCCAD_cad;
    if (cad?.schemaVersion !== 1 || cad.units !== "mm" || cad.coordinateSpace !== "PART_LOCAL" || !binary)
        throw new Error("Missing CAD GLB mapping");
    const read = (index: number, components: number, type: number): number[] => {
        const a = document.accessors?.[index], v = document.bufferViews?.[a?.bufferView];
        if (!a || !v || a.componentType !== type || a.type !== (components === 3 ? "VEC3" : "SCALAR") || v.buffer !== 0 || v.byteStride)
            throw new Error("Unsupported GLB accessor");
        const start = (v.byteOffset ?? 0) + (a.byteOffset ?? 0), length = a.count * components * 4;
        if (!Number.isSafeInteger(start) || !Number.isSafeInteger(v.byteLength) || v.byteLength < 0 || !Number.isSafeInteger(a.byteOffset ?? 0) || (a.byteOffset ?? 0) < 0 || !Number.isSafeInteger(a.count) || a.count < 0 || !Number.isSafeInteger(length) || start < 0 || length > (v.byteLength - (a.byteOffset ?? 0)) || start + length > binary!.byteLength)
            throw new Error("Invalid GLB accessor bounds");
        return Array.from({ length: a.count * components }, (_, i) => {
            const n = type === 5126 ? binary!.getFloat32(start + i * 4, true) : binary!.getUint32(start + i * 4, true);
            if (!Number.isFinite(n))
                throw new Error("Nonfinite GLB coordinate");
            return n;
        });
    };
    const vectors = (values: number[]): Vec3[] => Array.from({ length: values.length / 3 }, (_, i) => [values[i * 3], values[i * 3 + 1], values[i * 3 + 2]]);
    const primitives = document.meshes[0].primitives;
    if (primitives?.length !== 1)
        throw new Error("Invalid solid primitive count");
    mesh.vertices = vectors(read(primitives[0].attributes.POSITION, 3, 5126));
    const indices = read(primitives[0].indices, 1, 5125);
    if (indices.length % 3 || indices.some(id => id >= mesh.vertices.length))
        throw new Error("Invalid triangle index");
    mesh.triangles = vectors(indices);
    mesh.faceIds = read(cad.faceIds, 1, 5125);
    if (mesh.faceIds.length !== mesh.triangles.length)
        throw new Error("Invalid face mapping count");
    if (!Array.isArray(cad.edges) || !Array.isArray(cad.vertices))
        throw new Error("Invalid CAD topology mapping");
    const validId = (id: unknown) => typeof id === "number" && Number.isSafeInteger(id) && id > 0;
    mesh.edges = cad.edges.map((e: any) => { if (!validId(e.localId))
        throw new Error("Invalid CAD edge ID"); return { localId: e.localId, points: vectors(read(e.positions, 3, 5126)) }; });
    if (cad.vertices.some((v: any) => !validId(v.localId) || !Array.isArray(v.point) || v.point.length !== 3 || v.point.some((n: unknown) => typeof n !== "number" || !Number.isFinite(n))))
        throw new Error("Invalid CAD vertex mapping");
    mesh.topologyVertices = cad.vertices;
    mesh.stableIds = cad.stableIds ?? {};
    return { mesh, visualization };
}
// Mock/corpus authoring only. Production GLBs are emitted by the Geometry Worker.
export function encodeMeshGLB(mesh: MeshData, visualization?: VisualizationManifest): ArrayBuffer {
    const chunks: Uint8Array[] = [], views: any[] = [], accessors: any[] = [];
    let offset = 0;
    const add = (values: number[], components: number, float: boolean) => {
        const buffer = new ArrayBuffer(values.length * 4), view = new DataView(buffer);
        values.forEach((n, i) => float ? view.setFloat32(i * 4, n, true) : view.setUint32(i * 4, n, true));
        const index = accessors.length;
        views.push({ buffer: 0, byteOffset: offset, byteLength: buffer.byteLength });
        offset += buffer.byteLength;
        chunks.push(new Uint8Array(buffer));
        const bounds = float && components === 3 && values.length ? { min: [0, 1, 2].map(axis => Math.min(...values.filter((_, i) => i % 3 === axis))), max: [0, 1, 2].map(axis => Math.max(...values.filter((_, i) => i % 3 === axis))) } : {};
        accessors.push({ bufferView: index, componentType: float ? 5126 : 5125, count: values.length / components, type: components === 3 ? "VEC3" : "SCALAR", ...bounds });
        return index;
    };
    const position = add(mesh.vertices.flat(), 3, true), indices = add(mesh.triangles.flat(), 1, false), faceIds = add(mesh.faceIds, 1, false);
    const cad = { schemaVersion: 1, units: "mm", coordinateSpace: "PART_LOCAL", faceIds, edges: (mesh.edges ?? []).map(e => ({ localId: e.localId, positions: add(e.points.flat(), 3, true) })), vertices: mesh.topologyVertices ?? [], stableIds: mesh.stableIds ?? {} };
    const document = { asset: { version: "2.0" }, buffers: [{ byteLength: offset }], bufferViews: views, accessors, meshes: mesh.triangles.length ? [{ primitives: [{ attributes: { POSITION: position }, indices }] }] : [], nodes: [{ mesh: 0 }], scenes: [{ nodes: [0] }], scene: 0, extensionsUsed: ["OCCCCAD_cad", "OCCCCAD_visualization"], extensions: { OCCCCAD_cad: cad, OCCCCAD_visualization: visualization } };
    let json = new TextEncoder().encode(JSON.stringify(document));
    const padded = new Uint8Array(Math.ceil(json.length / 4) * 4).fill(32);
    padded.set(json);
    json = padded;
    const output = new ArrayBuffer(28 + json.length + offset), header = new DataView(output), bytes = new Uint8Array(output);
    header.setUint32(0, 0x46546c67, true);
    header.setUint32(4, 2, true);
    header.setUint32(8, output.byteLength, true);
    header.setUint32(12, json.length, true);
    header.setUint32(16, 0x4e4f534a, true);
    bytes.set(json, 20);
    header.setUint32(20 + json.length, offset, true);
    header.setUint32(24 + json.length, 0x004e4942, true);
    let cursor = 28 + json.length;
    for (const chunk of chunks) {
        bytes.set(chunk, cursor);
        cursor += chunk.length;
    }
    return output;
}
